proposals/0000-hybrid-output.md

```diff
@@ -0,0 +1,133 @@
+---
+
+<h1 set:html={text} />
+```
+
+# Motivation
+
+Since Astro v1.0.0 was released, we've gotten consistent feedback that more granular control over `output` is a requirement for scaling projects. Users would like to statically render some routes and server render others, reducing the need for full static rebuilds for frequently changing data.
+
+One very common use-case is a fully static site that also exposes `api/` endpoints which are handled on the server. Another common example is a largely static site that can generate a few dynamic pages at request time (for example, a user profile page).
+
+The eventual solution should enable granular control over `output` on a per-route basis. Since Astro defaults to `'static'` output, this should remain the default in the eventual solution as well.
+
+# Detailed design
+
+## Public API
+
+There are two public API surfaces to this RFC:
+
+- Updating the `astro.config.mjs` to accept a new `output` value, `'hybrid'`. The existing values of `'static' | 'server'` would remain supported. A value of `'hybrid'` will enabled detection for the following feature:
+- Allowing routes (files in `pages/`) to self-declare a granular `output` value. The existing `output` values of `'static' | 'server'` would both be valid.
+
+**For performance reasons, this declaration must be statically analyzable.**
+Only string literal values will be accepted. Other declaration values will throw an error and fail the build. This mirrors similar constraints that Vite has for compile-time (macro) features like dynamic `import()` and `import.meta.glob`.
+
+```astro
+---
+// Valid – constant string literal
+export const output = 'server'
+// Invalid – cannot be determined at compile-time
+export const output = static ? 'static' : 'server';
+export const output = value;
+export const output = 'ser' + 'ver';
+export let output = undefined; output = 'server';
+---
+```
+
+## Internal API
+
+Astro needs to detect/understand the `export const output` syntax. This should be implemented as a specific post-processing Babel plugin. The plugin will detect the existence of valid ESM exports of `output` (`export const output`, `export let output`, `export var output`, `export { output }`, etc) and attach the statically detected `output` value to the module metadata under the `astro` namespace. This will allow us to track the `output` mode of any public route.
+
+> **Note**
+> Astro Routes include `.astro` pages, but also API endpoints (`src/pages/api/user.ts`)! It's important that this RFC handles both of these constraints in a uniform way, hence the use of a post-processing Babel plugin rather than implementing this feature in `@astrojs/compiler`.
+
+The **build** API needs to be updated to handle the new `'hybrid'` output mode. Implementation-wise, this will be similar to the existing `'server'` implementation, but it will also adopt portions of the `'static'` build to statically generate any routes that are not handled at request time.
```

Comment:
It would be good to have a little more detail on implementation here. It's a little unclear to me how this will be broken apart in the build (although I have no doubt that it can be).

Reply:
Agreed, MVP is still in progress, but I'm hoping to flesh out this section a lot more in the coming days.